Ensure that ansible-runner worker --worker-info Returns Unique IDs

beeankha opened the pull request:
Fixes https://github.com/ansible/ansible-runner/issues/850

In certain cases (e.g. when utilizing VMs spun up with IBM Cloud), the provided machine ID for each node will sometimes be identical; this code change would make it so that we purposefully generate a brand new UUID for each node and store it in a file for retrieval (the details of this file and its location are still TBD); this will ensure an actually unique UUID for each worker node.

beeankha committed:
Generate a new UUID for each node vs discover assigned UUID, which could be duplicated across several nodes

beeankha commented:
recheck

samdoran commented on ansible_runner/utils/capacity.py:
```diff
@@ -21,17 +26,24 @@ def get_mem_in_bytes():
-            error = ("Could not find /var/lib/dbus/machine-id or "
-                     "/etc/machine-id files, UUID undiscoverable.")
-            return error
+def generate_or_get_uuid():
```
Functions that combine multiple capabilities like this are generally harder to test and maintain. It's better to have simple, single purpose functions. That allows the caller to control the desired behavior.

This function could be changed to call `get_uuid()`, then save it to disk. That way `get_uuid()` would be concerned with uuid generation and this function would be concerned with making sure the file exists on disk. Maybe call rename it to `save_uuid()`, have it accept a path to check or save to as a parameter, and return the value that it either gets from the path or save a new uuid to if the file doesn't exist.

samdoran commented on ansible_runner/utils/capacity.py:
```diff
@@ -1,5 +1,10 @@
+import glob
 import multiprocessing
+import os
```
You could use `pathlib` since this a Python 3 only project. That would remove the need for some of the other imports since `Path` objects have methods that provide the same functionality.

samdoran commented on ansible_runner/utils/capacity.py:
```diff
@@ -21,17 +26,24 @@ def get_mem_in_bytes():
         return error
 
 
-def get_uuid():
```
Rather than deleting a public function, why not fall back to `uuid.getnode()` or `uuid.uuid4()`?

samdoran replied:
Something like this may work:
```python
def get_uuid(path=''):
files = [
    Path('/var/lib/dbus/machine-id'),
    Path('/etc/machine-id'),
]

if path:
    files.append(Path(path))

for file in files:
    try:
        uuid = file.read_text()
        return uuid.strip()
     except FileNotFoundError:
        continue
        
return uuid.getnode()
```

samdoran commented on ansible_runner/utils/capacity.py:
```diff
@@ -21,17 +26,24 @@ def get_mem_in_bytes():
-                     "/etc/machine-id files, UUID undiscoverable.")
-            return error
+def generate_or_get_uuid():
+    uuid_directory = glob.glob('/tmp/node_uuid_*')  # Location TBD
```
I'm guessing this is a temporary placeholder. Using `glob` here could pick up multiple files, which could introduce a whole host of unexpected behavior. This should be looking at one and only one very specific location.

samdoran commented:
I just looked at the existing tests and there are _no_ tests at all for `get_uuid()`. That means that changing this function is risky since we 1) have nothing that validates the current behavior, 2) have no way of knowing if the changes made here break that behavior, and 3) have no tests for the new code.

At a minimum, we need tests for the net new code. It would be really great to create tests for the existing function before changing it as well.

beeankha committed:
Split UUID retrieval/generation into two separate functions

shanemcd commented on ansible_runner/utils/capacity.py:
```diff
@@ -22,16 +27,27 @@ def get_mem_in_bytes():
-            error = ("Could not find /var/lib/dbus/machine-id or "
-                     "/etc/machine-id files, UUID undiscoverable.")
-            return error
+    uuid_directory = Path('/etc/ansible/facts.d')
```
I'm curious why we picked this location. Is this a canonical place for this kind of file?

shanemcd replied:
Oh, I think I see why now. So that file will be read and sent back when the heartbeat runs? Neat.

chrismeyersfsu replied:
We needed a place. I picked a place that was ansibley. VERY open to a better place.

shanemcd replied:
Can someone point me to the place that we read this on the AWX side?

shanemcd commented on ansible_runner/utils/capacity.py:
```diff
@@ -22,16 +27,27 @@ def get_mem_in_bytes():
+    # Store the newly-generated UUID in a new dir/file
+    path = '/etc/ansible/facts.d'
+    uuid_file = 'uuid.txt'
+    uuid_file_path = pathlib.Path(path) / uuid_file
```
Is there any reason why this can't just be `os.path.join(path, uuid_file)`?

samdoran replied:
That would change the result from a `Path` object to a `str`.

shanemcd commented on ansible_runner/utils/capacity.py:
```diff
@@ -22,16 +27,27 @@ def get_mem_in_bytes():
+        return newly_generated_uuid
+
+
+def generate_uuid():
```
Can this function get marked private? I only see it being called internally in this module.

beeankha committed:
Look for uuid file specifically, instead of directory

chrismeyersfsu commented on ansible_runner/utils/capacity.py:
```diff
@@ -22,16 +27,28 @@ def get_mem_in_bytes():
+    uuid_file = 'uuid.txt'
+    uuid_file_path = uuid_dir / uuid_file
+
+    with uuid_file_path.open('w', encoding='utf-8') as uuid_file:
```
We want to make sure that the files starts its life with the correct permissions. Otherwise, an attack could write a program that just sits in a loop and keeps trying to "hijack" the new file that we are creating. They would be successful if their code ran in between the file open line and the chmod line.

Ansible core has to deal with this problem often. I think @samdoran can recommend the best way to deal with this.

samdoran replied:
What should the correct permissions be? How sensitive is this file?

The way Core (partially) defends against this type of race condition/attack is by the randomness of the parent temporary directory name, not the module payload file name. I'm not sure the payload that contains a module to be executed (in the case of Core) has the same risk profile as a file that contains a UUID. But I don't fully understand what a malicious actor could gain by substituting their own UUID for ours.

In the case of this file, we want it to be in a deterministic location, so randomizing the path and/or filename aren't really options. I suppose just setting the permissions to 400 would be enough?

That the default path is inside the user's home directory, and this code does not create any directories in the `uuid_file_path` also mitigates quite a bit of risk. The real trouble comes with `os.makedirs()` or `mkdir -p`, which isn't happening here.

samdoran replied:
In any case, it's probably a good idea to be explicit about the permissions on the file rather than relying on the `umask`. Should this be 600, 640, or mainly just not world readable?

samdoran replied:
I set the file to `600` by default, but the caller can change it.

beeankha committed:
Change location of UUID file to home dir

samdoran committed:
Change function names and signatures to reflect their behavior

Change both functions to accept a path to the UUID file, defaulting to an
immutable Path object. This allows easier testing without mocking.

Change function names to refelct that one sets the UUID (_set_uuid()), always
overwriting any existing file while the other function, ensure_uuid(), will
only set the UUID if the UUID file does not exist.

Add unit tests.

samdoran commented on ansible_runner/utils/capacity.py:
```diff
@@ -24,24 +24,21 @@ def get_mem_in_bytes():
         return error
 
 
-def get_uuid():
```
Do we need to keep `get_uuid()` around for backwards compatibility?

samdoran replied:
It probably wouldn't hurt. We can just deprecate it. Deleting it entirely is a bit abrupt since it's a public function.

shanemcd replied:
I don't think so, it was never shipped, just landed a month ago https://github.com/ansible/ansible-runner/commit/49353d2d1d6a4fbb9ed47c3ca9a2b6e82f392e41

samdoran replied:
Sweet!

beeankha committed:
Import and call ensure_uuid() function by updated name

samdoran committed:
Use a simpler and consistent default value for uuid path

While Path objects are immutable, the previous default had two main problems:

  1. Function defaults are processed on import. The previous default meant that
     quite a bit of work was done just when the function was imported.
  2. Since the value of Path.home() varies depending on the user invoking the
     function, the means the function default is not consistent.

Using None then setting the default value within the function is a bit noisier,
but addresses the two issues above.

samdoran committed:
Ensure the mode of the UUID file is set properly

Add parameters for accepting mode.
Add tests for ensuring the mode is set properly.

samdoran committed:
Fix bug in call to _set_uuid()

samdoran committed:
Set UUID file permissions correctly even if the file already exists

samdoran commented:
recheck
